### PR TITLE
Retain instrumentation after adding event to batch

### PR DESCRIPTION
**Pull request by JoshLove-msft**
Addresses https://github.com/Azure/azure-sdk-for-net/issues/34763

### CODE REVIEW

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/EventDataBatch.cs [hunk 1]
```diff
@@ -180,34 +180,16 @@ public bool TryAdd(EventData eventData)
...
-                {
-                    ClientDiagnostics.InstrumentMessage(eventData.Properties, DiagnosticProperty.EventActivityName, out messageScopeCreated, out traceparent, out var tracestate);
-
-                    var added = InnerBatch.TryAdd(eventData);
-
-                    if ((added) && (traceparent != null))
-                    {
-                        EventDiagnosticIdentifiers.Add((traceparent, tracestate));
-                    }
+                var added = InnerBatch.TryAdd(eventData);
```
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/EventDataBatch.cs:185]: If we do not accept the event into the batch, we should continue to remove the instrumentation, no?   Why would we keep it?

**Review by jsquire — CHANGES_REQUESTED** [review]

**Comment by JoshLove-msft** [comment]
> It just feels a bit wrong to me that we're modifying data that you give us for a `TryAdd` - even when we explicitly reject it.

Yeah I would agree here - it probably makes more sense to remove the identifier if it gets rejected.

**Comment by m-redding** [comment]
I feel like since `SendAsync` doesn't remove the instrumentation on failure, the best parallel for batches is to also not remove instrumentation on failure? Even though the failures have different natures, it could potentially be useful to have tracing information for all failures. 
The alternative could be to instrument the whole batch when it is passed to send, rather than when events are added to the batch. But I'm not sure about the performance implications/implementation challenges of that.

**Comment by lmolkova** [comment]
I agree that adding context for rejected events is not helpful. I wonder if it's possible to instrument the message only after it's successfully added to the batch?

We instrument individual events so they
- capture the context in which they were added to batch
- we can add link from send span to all messages being published. So we already have span per publish and span per each message. 

If message is added to batch, but never published, users would see it (there will be no publish span).
If message was never added, there is no immediate need to trace this fact.

**Comment by JoshLove-msft** [comment]
> If message is added to batch, but never published, users would see it (there will be no publish span).
> If message was never added, there is no immediate need to trace this fact.

The way things are currently implemented, we create the Message/Event scope regardless of whether the message/event  is successfully added to the batch. I'm not actually sure if we can avoid doing this if we want to instrument the message/event up front so that the calculation is correct (we could maybe just add overhead for the diagnostic-id without instrumenting up front). Assuming that we stick with the current implementation of creating the message/event scope up-front, wouldn't it be more confusing for users if this scope is not reflected anywhere in the message/event properties?

**Comment by jsquire** [comment]
> The way things are currently implemented, we create the Message/Event scope regardless of whether the message/event is successfully added to the batch. I'm not actually sure if we can avoid doing this if we want to instrument the message/event up front so that the calculation is correct (we could maybe just add overhead for the diagnostic-id without instrumenting up front). Assuming that we stick with the current implementation of creating the message/event scope up-front, wouldn't it be more confusing for users if this scope is not reflected anywhere in the message/event properties, but they still see that a span was created in the telemetry?

If we do that, I think we end up with a non-trivial performance impact as we would have to serialize twice - once to measure and once with the actual instrumentation.

What we do today is:
- Instrument the message
- Serialize to AMQP format and measure
- If we accept the message in the batch, we keep the serialized bytes

If we were to instrument with a dummy to measure, we would have to:
- Instrument the message with the dummy value
- Serialize to AMQP format and measure
- If we accept the message in the batch, then instrument it for real, serialize it to AMQP format again, keep the serialized bytes



What happens if we create the scope, instrument the event, we reject it, then send the batch... then we add that event to a new batch and send that.  What would someone see in telemetry?   Does that make it look like the event was in two batches when it was really only in one of them?

**Comment by lmolkova** [comment]
Chatted with @JoshLove-msft offline and decided that injecting context regardless is a good feature.

Rationale:
The context we capture on the event should continue the one when the message was created (for correlation and causation). E.g. when I get incoming requests from the upstream service, I create an event and add it to a batch. The event's context should be a child of the incoming request.
If the event does not fit into a batch and is published later on in background thread or in scope of some other operation, it should not change the context.

So:
1. I created a message and tried to add it to a batch at 1:58 pm 
2. It did not fit and I  added it to a different one at 2:00 pm
3. it was published at 2:00:01 pm

 
It's cool to see that message was sitting and waiting to be added to batch for 2 minutes. Otherwise I'd see

1. something appeared out of blue at 2:00 pm, it has no parent (e.g. stayed in the background queue)
2. it was published at 2:00:01 pm

To provide enough observability into this we can add an attribute explaining event was not added right away:

```csharp
if (hasTraceContext() {
  using var scope = createMessageScope(msg)
  if (!tryAdd(msg)) {
     scope.SetAttribute("message.added", false); // need to think about naming a bit
 }
} else { tryAdd(msg) }
```

This way I can also write a query to check if any messages were dropped because adding was not successful.

**Comment by JoshLove-msft** [comment]
> Chatted with @JoshLove-msft offline and decided that injecting context regardless is a good feature.
> 
> Rationale: The context we capture on the event should continue the one when the message was created (for correlation and causation). E.g. when I get incoming requests from the upstream service, I create an event and add it to a batch. The event's context should be a child of the incoming request. If the event does not fit into a batch and is published later on in background thread or in scope of some other operation, it should not change the context.
> 
> So:
> 
> 1. I created a message and tried to add it to a batch at 1:58 pm
> 2. It did not fit and I  added it to a different one at 2:00 pm
> 3. it was published at 2:00:01 pm
> 
>   It's cool to see that message was sitting and waiting to be added to batch for 2 minutes. Otherwise I'd see
> 
> 1. something appeared out of blue at 2:00 pm, it has no parent (e.g. stayed in the background queue)
> 2. it was published at 2:00:01 pm
> 
> To provide enough observability into this we can add an attribute explaining event was not added right away:
> 
> ```cs
> if (!hasTraceContext(msg)) {
>   using var scope = createMessageScope(msg)
>   if (!tryAdd(msg)) {
>      scope.SetAttribute("message.added", false); // need to think about naming a bit
>  }
> } else { tryAdd(msg) }
> ```
> 
> This way I can also write a query to check if any messages were dropped because adding was not successful and the application didn't check.

This also aligns with @m-redding's reasoning!

**Comment by JoshLove-msft** [comment]
Should we proceed with the PR as-is and come back later to include "message.added" (maybe it needs to be talked over with Otel folks?), or do we think it is important to have this attribute up front since we are changing the behavior in Event Hubs?

**Comment by lmolkova** [comment]
> Should we proceed with the PR as-is and come back later to include "message.added" (maybe it needs to be talked over with Otel folks?), or do we think it is important to have this attribute up front since we are changing the behavior in Event Hubs?

I'm fine with proceeding as is and adding attribute later.

**Review by jsquire — APPROVED** [review]
per discussion, it looks like we landed on "we should instrument even when the event is not added" with the rationale that we'll have traceability on that activity and, ultimately, when it was sent.

**Pull request merged**